aten/src/ATen/BatchedTensorImpl.cpp

```diff
@@ -0,0 +1,95 @@
+#include <ATen/BatchedTensorImpl.h>
+
+#include <ATen/WrapDimUtils.h>
+#include <c10/util/Exception.h>
+
+namespace at {
+
+BatchedTensorImpl::BatchedTensorImpl(Tensor value, BatchDims bdims)
+  : TensorImpl(
+      c10::DispatchKeySet(DispatchKey::Vmap),
+      value.dtype(),
+      value.device()
+    )
+  , value_(std::move(value))
+  , bdims_(std::move(bdims))
+{
+  TORCH_INTERNAL_ASSERT(value_.defined());
+
+  const auto public_dims = value_.dim() - bdims_.size();
+  const auto value_sizes = value_.sizes();
+  sizes_.clear();
+  sizes_.reserve(public_dims);
+  for (int64_t dim = 0; dim < public_dims; dim++) {
+    auto actual_dim = actualDim(dim, /*wrap_dim=*/false);
+    sizes_.push_back(value_sizes.at(actual_dim));
+  }
+  refresh_numel();
+}
+
+int64_t BatchedTensorImpl::actualDim(int64_t dim, bool wrap_dim) const {
+  if (wrap_dim) {
+    const auto ndim = sizes_.size();
+    dim = maybe_wrap_dim(dim, ndim);
+  }
+
+  auto is_bdim = createIsBdimBitset(bdims_);
+  int64_t non_bdim_count = 0;
+  for (int64_t actual_dim = 0; actual_dim < kVmapMaxTensorDims; actual_dim++) {
+    if (is_bdim[actual_dim]) {
+      continue;
+    }
+    if (non_bdim_count == dim) {
+      return actual_dim;
+    }
+    non_bdim_count++;
+  }
+  TORCH_INTERNAL_ASSERT(false);
```

Comment:
Can you, please, add a comment on why we shouldn't get here? Or add an error message that would be obvious.